### PR TITLE
Enable building of DevTools and scheduling profiler in CI

### PR DESCRIPTION
**Pull request by taneliang**
## Summary

This PR makes 4 separate changes to the CI config:

1. Re-enables building of main DevTools in CI. This was previously disabled in #19012. I'm not sure why builds were failing, but I'm guessing it's due to the build issue that #19647 fixed.

1. I've disabled parallelism in the DevTools build job because it looked like it was just doing the same thing 20 times across 20 executors 😆 I don't think this job is worth parallelizing as it isn't the longest one, but I can look into parallelizing this if we really wanted to.

1. #19566 replaced `yarn install` commands in every job with a central cache of the root `node_modules` folder. However, the `node_modules` folders in each package were not cached. Because `yarn install` was no longer run in each job, these `node_modules` folders were no longer generated.
	
	If we SSH into a CI executor after the setup job:

	```sh
	circleci@a41afdc807a2:~/project$ ls packages/react-devtools-scheduling-profiler
	README.md  package.json  src  webpack.config.js
	circleci@a41afdc807a2:~/project$ ls packages/*/node_modules
	ls: cannot access 'packages/*/node_modules': No such file or directory
	```

	This is a truncated list of the contents of all project-specific `node_modules` folders on my dev computer:

	```sh
    $ ls packages/*/node_modules
    packages/eslint-plugin-react-hooks/node_modules:

    packages/jest-mock-scheduler/node_modules:

    packages/jest-react/node_modules:

    packages/react-art/node_modules:

    packages/react-client/node_modules:

    packages/react-devtools-core/node_modules:
    ws/

    packages/react-devtools-extensions/node_modules:
    ansi-styles/      browserify-zlib/        crc32-stream/        ...

    packages/react-devtools-inline/node_modules:
    ansi-styles/  '@babel'/   babel-eslint/   chalk/   ...

    packages/react-devtools/node_modules:

    packages/react-devtools-scheduling-profiler/node_modules:
    acorn/               css-loader/            loader-utils/                       ...

    packages/react-devtools-shared/node_modules:
    semver/

    packages/react-devtools-shell/node_modules:
    ansi-styles/  '@babel'/   babel-eslint/   chalk/   ...

    packages/react-dom/node_modules:

    packages/react-interactions/node_modules:

    packages/react-native-renderer/node_modules:
    scheduler/

    packages/react/node_modules:

    packages/react-reconciler/node_modules:

    packages/react-server/node_modules:

    packages/react-transport-dom-relay/node_modules:
    scheduler/

    packages/react-transport-dom-webpack/node_modules:

    packages/scheduler/node_modules:

	```

	Because of this, the scheduling profiler could not be built in CI. ~This PR adds paths to all packages' `node_modules` folders (even if they don't exist) to the CI cache.~ This PR calls `yarn install` during the DevTools build job as suggested by https://github.com/facebook/react/pull/19691#issuecomment-686709981. More details in this discussion thread: https://github.com/facebook/react/pull/19691#discussion_r482974891.

1. Adds the scheduling profiler to the DevTools build script. We can eventually add another CI job to deploy the artifact to Vercel. Question (to @bvaughn): should I also compress the `scheduling-profiler` folder in the devtools.tgz artifact? **UPDATE:** As suggested in https://github.com/facebook/react/pull/19691#issuecomment-686486724, the scheduling profiler build is stored in a new `devtools-scheduling-profiler.tgz` artifact.

## Test Plan

1. CI
1. Inspecting the devtools-scheduling-profiler.tgz [artifact](https://app.circleci.com/pipelines/github/facebook/react/5033/workflows/c99adec9-5cfd-4036-b8c5-131b2fd28688/jobs/210722/artifacts) and serving the scheduling-profiler.

```sh
$ tar -xvzf devtools-scheduling-profiler.tgz
./
./app.js
./import.worker.worker.js.map
./static/
./static/media/
./static/media/profilerBrowser.8d782287.png
./import.worker.worker.js
./index.html
./app.js.map
```

### CODE REVIEW

File: .circleci/config.yml
Comment (taneliang) [inline .circleci/config.yml:84]: This is the specific error we get if we put everything in one cache: https://support.circleci.com/hc/en-us/articles/360037695934-save-cache-step-fails-with-Error-uploading-archive-MetadataTooLarge-Your-metadata-headers-exceed-the-maximum-allowed-metadata-size-

**Comment by bvaughn** [comment]
This is in my queue to review. I was on PTO last week though and I have a lot of catch up to do. I'll try to get to it sometime in the next couple of days. :smile:

**Comment by bvaughn** [comment]
> Question (to @bvaughn): should I also compress the `scheduling-profiler` folder in the `devtools.tgz` artifact?

Can we actually put this into a separate TGZ file? `devtools.tgz` is for things being deployed to Chrome/Firefox/Edge stores. This new thing gets deployed elsewhere, during a different workflow.

File: .circleci/config.yml
Comment (bvaughn) [inline .circleci/config.yml:24]: cc @rickhanlonii for a second set of eyes on the Circle CI changes.

File: .circleci/config.yml
Comment (bvaughn) [inline .circleci/config.yml:136]: This is something that will be easy to forget to update when we next add a new package. Will the (CI) failure behavior be obvious? (Will the error message reasonably direct folks to this bit of the config?)

File: .circleci/config.yml
Comment (bvaughn) [inline .circleci/config.yml:456]: I don't think we need to restore DevTools modules for most of these CI tasks, do we?

Just `build_devtools_and_process_artifacts` and `yarn_test_build_devtools`?

(And maybe we don't need to `restore_package_node_modules` for those two tasks either?)

File: .circleci/config.yml
Comment (bvaughn) [reply .circleci/config.yml:136]: That sounds more robust.

File: .circleci/config.yml
Comment (taneliang) [reply .circleci/config.yml:456]: Hmm, I wasn't sure what was and wasn't needed by the various CI tasks, so I figured it'll be safest to just restore everything. I'd expect e.g. the build, test, flow tasks to need everything restored?

Either way, the devtools and package caches are really cheap to restore (~0s) so maybe it won't be worth checking if they're necessary?

image


In any case, if we use the consolidated node_modules cache approach in my comment above we can go back to having a single cache :D

**Comment by taneliang** [comment]
> Can we actually put this into a separate TGZ file? devtools.tgz is for things being deployed to Chrome/Firefox/Edge stores. This new thing gets deployed elsewhere, during a different workflow.

Okay! I'll do that tomorrow.

Regarding deployment, do you have any workflow for that in mind? Seems like we won't be able to integrate Vercel's bot normally because it doesn't have Java and so it can't build React. The best idea I've come up with so far is this: write a script to upload the compiled app to Vercel, and run that in a master-only CI job. Wondering if you have other ideas 😆

**Comment by bvaughn** [comment]
> Regarding deployment, do you have any workflow for that in mind? 

Haven't had the time to give it a lot of thought yet. Open for suggestions!

File: .circleci/config.yml
Comment (rickhanlonii) [reply .circleci/config.yml:136]: Can we run a yarn install for only the tests that can't use the global cache? Is that only the devtools test?

**Comment by rickhanlonii** [comment]
@taneliang instead of caching all the package node modules, could we run a yarn install for only the new test? The install should be fast enough because we're caching the global yarn cache, we just don't want to do it in the other jobs because of the parallelization.

File: .circleci/config.yml
Comment (taneliang) [reply .circleci/config.yml:136]: Yeah we could do that. I'm just concerned that these missing folders may also cause problems in the future, so it would seem less risky to me if we just cached all node_modules folders. `react-native-renderer` and `react-transport-dom-relay` currently do have non-empty node_modules too (even though they only contain `scheduler`).

What do you think? I'll be happy to change this PR to either approach.

Comment (bvaughn) [reply .circleci/config.yml:136]: Will defer to @rickhanlonii since he's more familiar with this than I am.

Comment (taneliang) [reply .circleci/config.yml:136]: > If we ever have 2+ nested deps that are different versions, this consolidated approach would cause problems.

Hmm, I was thinking of copying all the packages' node_modules folders into individual folders. So the cache will contain multiple `package_node_modules/<package name>/node_modules` folders alongside the root `node_modules` folder. This will allow us to restore all the packages' `node_modules` folders back to where `yarn install` created them. I think this should be pretty safe, if a little complicated.

My concern with running `yarn install` only for the new test is that other tests may suffer from silent problems if their conflicting deps aren't restored/installed. The set of conflicting deps may change over time as well.

Comment (bvaughn) [reply .circleci/config.yml:136]: Sorry then. I misunderstood what you were suggesting. That sounds fine. 👍

Comment (rickhanlonii) [reply .circleci/config.yml:136]: >  I think the consolidated cache idea seems problematic if we ever have conflicts between packages

Yarn workspaces hoist all of the dependencies in a package up to the root node_modules already unless there's a conflict. So to prevent conflicts, we could enforce all packages use the same version and add a test to compare the package.json files. What do you think about that?

Comment (bvaughn) [reply .circleci/config.yml:136]: We already have several conflicts. Some of them probably arbitrary due to transient deps, others perhaps intentional (like `react-native-renderer` + `scheduler`).

Comment (rickhanlonii) [reply .circleci/config.yml:136]: Btw this is what we do internally in xplat/js.

**Comment by bvaughn** [comment]
Did this stall out or is it waiting on another review? Just noticed it had been a few days since anyone commented on it.

**Comment by taneliang** [comment]
Oops! Been swamped by schoolwork.

I'm also not sure how I should proceed. Should I do what @rickhanlonii originally suggested, I.e. remove the new caches and run `yarn install` in the devtools jobs? We can add that package.json test to prevent conflicting deps in a future PR.

**Comment by bvaughn** [comment]
No worries @taneliang.

> Should I do what @rickhanlonii originally suggested

I don't have any objections to that.

**Comment by taneliang** [comment]
Sounds good! I'll do this over the coming weekend.

**Comment by taneliang** [comment]
All done!

Unfortunately it looks like this `build_devtools_and_process_artifacts` job is now the longest one. I'm not sure what we should do about this. The bulk of the time is spent building main DevTools stuff (see [logs](https://app.circleci.com/pipelines/github/facebook/react/5033/workflows/c99adec9-5cfd-4036-b8c5-131b2fd28688/jobs/210722)), so moving the scheduling profiler build into a separate job won't make it *that* much faster but can probably reduce the total time by ~20-30s. Should I do that?

image

**Comment by taneliang** [comment]
Actually, I'll split the scheduling profiler build into a separate job, if everyone's okay with that.

I intend to add a deploy job in a future PR, and it'll be dependent on the build job. Although we don't *have* to split the build job, it'll allow the deployment to be done in parallel with the main DevTools build.

This is what the final job tree will look like:

```
... -> yarn_build -> build_devtools_scheduling_profiler -> deploy_devtools_scheduling_profiler (I'll add this in a future PR, and it'll only be run on master)
```

**Comment by bvaughn** [comment]
@taneliang I'm looking at the most recent CI run for this branch:
https://app.circleci.com/pipelines/github/taneliang/react?branch=scheduling-profiler-ci

Looks like it just stopped running things midway through?
Screen Shot 2020-09-21 at 9 36 33 AM

Am I misreading things maybe?

**Comment by taneliang** [comment]
Interesting. Looks like CircleCI is showing the CI runs from the project on my CircleCI account instead of Facebook's. Most of the jobs on my account aren't running because my free account doesn't have 20x parallelism.

Let me finish working on my deployment PR and I'll deactivate the project on my account, which should hopefully fix this.

**Comment by bvaughn** [comment]
Oh that is interesting.

Okay! :)

**Comment by taneliang** [comment]
Disabled the project on my CircleCI account. Closing and reopening to kick CI.

**Comment by taneliang** [comment]
@bvaughn this is ready for review!

image

**Comment by bvaughn** [comment]
Excellent! Thanks 🙇  Reviewing now.

File: .circleci/config.yml
Comment (bvaughn) [inline .circleci/config.yml:302]: Shouldn't this be tagged with `RELEASE_CHANNEL: experimental`?

The scheduling profiler uses `unstable_createRoot`. The stable `yarn_build` command doesn't build a version of ReactDOM that exposes this API.

Comment (taneliang) [reply .circleci/config.yml:302]: I don't think so, unless I'm missing something. The `yarn_build` CI job builds everything with the experimental channel. Because we depend on those packages, I don't think the scheduling profiler's build will be affected by a `RELEASE_CHANNEL` tag.

I've also confirmed that the built artifact works:

image


image

Comment (bvaughn) [reply .circleci/config.yml:302]: Oh, huh. I guess I'm a little out of date on our how Circle CI config is setup. `build_devtools_and_process_artifacts` specifies `RELEASE_CHANNEL: experimental`, but maybe it's unnecessary since, as you say, `yarn_build` is in the experimental channel (and `RELEASE_CHANNEL_stable_yarn_build` is stable)

**Review by bvaughn — APPROVED** [review]

File: .circleci/config.yml
Comment (taneliang) [reply .circleci/config.yml:302]: Yeah, there's a good chance that that's unnecessary, but I didn't investigate that as it was already there 😆

**Pull request merged**
